Quick Starts for Console (#360)
openshift/enhancements: enhancements/console/quick-starts.md

### Goals

1. Provide a mechanism to display user guides for various workflows.
2. Make users understand what are the steps needed to achieve their goals.
3. Provide new CRD format for writing the Quick Starts.
4. Have a repository for the out-of-the-box Quick Starts, those that describe how to install an operator or go though standard workflow.

## Proposal

* Introduce CRD format that will be used for writing Quick Starts.
* Introduce default repository for Quick Starts.
  
### User Stories

#### Story 1

As an administrator of an OpenShift cluster, I need a guide to walk me through how to install OpenShift Serverless modules (Serving and Eventing) in a cluster.

#### Story 2

As an administrator of an OpenShift cluster, I need a guide to walk me through how to update an OpenShift cluster.

#### Story 3

As a developer, I need a guide to walk me through how to deploy an existing application as a serverless workload.

#### Story 4

As a operator creator I want to provide operator consumers with a guide on how to install and user the my operator.

### Implementation Details

1. In order to provide mechanism to discribe a Quick Start, new CRD named `QuickStarts` will be created.

> 2. A new `openshift/quick-tours` repository will be created which will contain all supported Quick Starts. The `openshift/quick-tours` repository will have `release-*` branches.

Reviewer: `openshift/quick-starts`?

3. `console-operator` will import all the existing Quick Starts CRs from appropriate branch of the `openshift/quick-starts` repository into the `/manifest` directory, so that the CVO can:
   * create them if the CR doesn't exists.
   * update the CRs upon the cluster update (since CVO is doing `apply`).

> 4. `console-operator` will be resposible for removing older `QuickStarts` versions, based on their CRs `spec.version` field that will be in semver format. If the semver version range won't match the cluster version the `console-operator` will remove the CR.

Reviewer: Is this necessarily? It seems like we can handle versioning by having release-* branches in the quick-starts repository and just vendoring the right version. I'm not sure we need a version range in the quick start resource itself.

Author: So field was mainly added, so the operator can get rid of old QuickStarts, since after the cluster update they wont be deleted from the cluster.
Also I think it cant hurt, for the QuickStart CR to know whats it's version.

Reviewer: Hey @jhadvig. I'm still unclear. The new quick start for the current version should just replace the old one since they have the same name. In that case, the operator doesn't need to do anything.

If we need to remove a quick start entirely, I think we might have to hard-code that special case in the operator for one release. I wouldn't expect this to happen often.

Author: ok! makes sense, will remove.

      review:
        instructions: Make sure the knative-serving API was successfully installed ...
        taskHelp: Try walking through the steps again to properly create the instance of knative-serving
      recapitulation:
        success: You've just created an instance of knative-serving! Next, we'll create an instance of knative-eventing
        failed: Check your work to make sure that the instance of knative-serving is properly created
    - title: create knative-eventing API
      description: The second CR we'll create is knative-eventing ...
      review:
        instructions: Make sure the knative-eventing API was successfully installed ...
        taskHelp: Try walking through the steps again to properly create the instance of knative-eventing
      recapitulation:
        success: You've just created an instance of knative-eventing!
        failed: Check your work to make sure that the instance of knative-eventing is properly created
  conclusion: Your Serverless Operator is ready! If you want to learn how to deploy a serverless application, take the Serverless Application tour.
  nextQuickStart: serverless-application
```

#### Quick Start icon

> Each Quick Start CR can contain an icon that is specified in the `spec.icon` field. This field is either an URL of the Package Manifest that contains icon, or a base64 encoded image that are used as Quick Start's icon

Reviewer: I don't think we should ask quick start authors to build a URL to a package manifest icon. That's error prone and is really an internal implementation detail that could change over time.

Author: Wasn't sure with it either but that's it it the outcome of the conversation I had vit @rohitkrai03 and @christianvogt.
Do you propose to cover just the base64 encoded image ?

Reviewer: If we want to support this, I'd suggest a specific API for referencing an operator icon by package manifest name.

#### Quick Start's Access Review

Since different users might not have access to specific actions on different resources, it would be pointless for them to go though a Quick Start and be unable to complete its tasks. For that reason the `spec` of the Quick Start CRD contains also `accessReviewResources` field, that should contain an array of resources and actions done on them, that user should perform during taking the Quick Start tour. User's access to the Quick Start is reviewed
based on array of resource actions. In order for the user to see the Quick Start and take its tour, his access review needs to pass all the listed resources actions.

#### Air Gapped Environments

Since the supported Quick Starts CRs will be part of the `console-operator`'s manifests, they will be distributed together with image.

### Third party Quick Starts

> Third party Quick Starts will need to be created manually on the cluster, or upon operator installation.

Reviewer: I'd make it clear that it's the operator itself that creates these after installation.

Author: How about?

```
### Installation

`console-operator` is responsible for installing supported Quick Starts at the end of the cluster provisioning process. Third party Quick Starts will need to be created manually on the cluster.
```

Reviewer: I was thinking something like

```suggestion
Third party Quick Starts will need to be created by an operator after it's installed or manually by a cluster administrator.
```
